src/System.Private.ServiceModel/src/System/ServiceModel/Security/WSSecurityOneDotZeroSendSecurityHeader.cs

```diff
@@ -2,20 +2,819 @@
+                }
+            }
+            else
+            {
+                AsymmetricSecurityKey asymmetricKey = signatureKey as AsymmetricSecurityKey;
+                if (asymmetricKey == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(
+                        SR.Format(SR.UnknownICryptoType, _signingKey)));
+                }
+
+                asymmetricAlgorithm = asymmetricKey.GetAsymmetricAlgorithm(algorithmName, privateKey: true);
+                if (asymmetricAlgorithm == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(
+                        SR.Format(SR.UnableToCreateHashAlgorithmFromAsymmetricCrypto, algorithmName,
+                            asymmetricKey)));
+                }
+            }
+        }
+
+        protected override ISignatureValueSecurityElement CreateSupportingSignature(SecurityToken token, SecurityKeyIdentifier identifier)
+        {
+            StartPrimarySignatureCore(token, identifier, MessagePartSpecification.NoParts, false);
+            return CompletePrimarySignatureCore(null, null, null, null, false);
+        }
+
+        protected override ISignatureValueSecurityElement CreateSupportingSignature(SecurityToken token, SecurityKeyIdentifier identifier, ISecurityElement elementToSign)
+        {
+            string signatureAlgorithm;
+            XmlDictionaryString signatureAlgorithmDictionaryString;
+            SecurityKey signatureKey;
+            AlgorithmSuite.GetSignatureAlgorithmAndKey(token, out signatureAlgorithm, out signatureKey, out signatureAlgorithmDictionaryString);
+
+            System.Security.Cryptography.Xml.SignedXml signedXml = new SignedXml();
+            System.Security.Cryptography.Xml.SignedInfo signedInfo = signedXml.SignedInfo;
+            signedInfo.CanonicalizationMethod = AlgorithmSuite.DefaultCanonicalizationAlgorithm;
+            signedInfo.SignatureMethod = signatureAlgorithm;
+
+            if (elementToSign.Id == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.ElementToSignMustHaveId));
+            }
+
+            MemoryStream stream = new MemoryStream();
+            XmlDictionaryWriter utf8Writer = TakeUtf8Writer();
+            utf8Writer.StartCanonicalization(stream, false, null);
+            elementToSign.WriteTo(utf8Writer, ServiceModelDictionaryManager.Instance);
+            utf8Writer.EndCanonicalization();
+            stream.Position = 0;
+            var reference = new System.Security.Cryptography.Xml.Reference(stream);
+            reference.Uri = "#" + elementToSign.Id;
+            reference.DigestMethod = this.AlgorithmSuite.DefaultDigestAlgorithm;
+            reference.AddTransform(new XmlDsigExcC14NTransform());
+            _signedXml.AddReference(reference);
```

Comment:
It really feels like perhaps you want a helper like

```C#
private void AddReference(string id, MemoryStream contents)
{
    contents.Position = 0;
    ...
    _signedXml.AddReference(reference);
}
```

because you seem to have a lot of copies of these 5 lines of code.